### PR TITLE
Issue/none introduce connection state

### PR DESCRIPTION
**Pull request by wyrzyk**

### CODE REVIEW

File: src/main/java/com/atlassian/db/replica/api/state/States.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.atlassian.db.replica.api.state;
+
+public enum States {
```
Comment (wyrzyk) [inline src/main/java/com/atlassian/db/replica/api/state/States.java:3]: I probably should not use enums for API.

Comment (dagguh) [reply src/main/java/com/atlassian/db/replica/api/state/States.java:3]: Hmm, but transitions would be tricky: `getReadConnection()` would have to return both a `Connection` and a new `ConnectionState`. Or become mutable.

File: src/main/java/com/atlassian/db/replica/spi/state/StateListener.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package com.atlassian.db.replica.spi.state;
+
+import com.atlassian.db.replica.api.state.States;
+
+public interface StateListener {
```
Comment (dagguh) [inline src/main/java/com/atlassian/db/replica/spi/state/StateListener.java:5]: So far it seems it only has a no-op impl. What's the target use case?

File: src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        if (isClosed != null && isClosed) {
+            return CLOSED;
+        } else if (!readConnection.isInitialized() && !writeConnection.isInitialized()) {
+            return NOT_INITIALISED;
+        } else if (writeConnection.isInitialized()) {
+            return MAIN;
+        } else if (readConnection.isInitialized()) {
+            return REPLICA;
+        } else {
+            throw new UnsupportedStateException();
```
[the remark this answers is not in the document]
Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java:69]: The whole class and DualConnection is not thread-safe ;)

File: src/main/java/com/atlassian/db/replica/spi/state/StateListener.java (hunk 2 above)
Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/spi/state/StateListener.java:5]: 1. It's easier to test how states changes.
2. We can track transitions more gradually (for example, with the stack-trace mechanism) form the app.

I'm not 100% sure if we need this extra visibility available in API.

File: src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dagguh) [reply src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java:69]: Yeah but it didn't throw `UnsupportedStateException`s earlier.
Perhaps the race condition is bogus, if we have the guarantee that `isInitialized` cannot go from `true` to `false`.

File: src/main/java/com/atlassian/db/replica/api/state/States.java (hunk 1 above)
Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/api/state/States.java:3]: What do you mean? Can you express it with a test case?

File: src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java:69]: There's no guarantee. 

- [ ] Avoid `UnsupportedStateExceptions`

[the remark this answers is not in the document]
Comment (dagguh) [reply src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java:69]: Right, we have the `.reset`

File: src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java [hunk 4]
```diff
@@ -1,104 +1,75 @@
...
-        initializedConnections.clear();
-        if (!autoCommitBefore.equals(getAutoCommit())) {
+    public void setAutoCommit(Boolean autoCommit) throws SQLException {
+        final boolean autoCommitBefore = getAutoCommit();
+        parameters.setAutoCommit(autoCommit);
+        final Optional<Connection> connection = state.getConnection();
+        if (connection.isPresent()) {
+            connection.get().setAutoCommit(autoCommit);
+        }
+        if (autoCommitBefore != getAutoCommit()) {
```
Comment (dwatl) [inline src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java:62]: autoCommitPrev(ious)?

**Review by dwatl — APPROVED** [review]
LGTM

File: src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java (hunk 4 above)
Comment (wyrzyk) [reply src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java:62]: Can you elaborate?

Comment (dwatl) [reply src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java:62]: I meant the naming, but it's highly subjective - **previous** seemed more clear than **before**, but I'm not sure now

File: src/main/java/com/atlassian/db/replica/api/state/State.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    private final String name;
+
+    private State(String name) {
+        this.name = name;
+    }
+
+    public static final State NOT_INITIALISED = new State("NOT_INITIALISED");
+    public static final State MAIN = new State("MAIN");
+    public static final State REPLICA = new State("REPLICA");
+    public static final State CLOSED = new State("CLOSED");
```
[the remark this answers is not in the document]
Comment (dagguh) [reply src/main/java/com/atlassian/db/replica/api/state/State.java:16]: Oops, I forgot to submit the review 😅 
Either way, merging is a good call. Due to VPN problems I didn't find time to prepare the example, so better merge than wait 👍

File: src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java [hunk 6]
```diff
@@ -1,104 +1,75 @@
...
-        final Boolean autoCommitBefore = getAutoCommit();
-        this.isAutoCommit = autoCommit;
-        initializedConnections.clear();
-        if (!autoCommitBefore.equals(getAutoCommit())) {
+    public void setAutoCommit(Boolean autoCommit) throws SQLException {
+        final boolean autoCommitBefore = getAutoCommit();
+        parameters.setAutoCommit(autoCommit);
+        final Optional<Connection> connection = state.getConnection();
+        if (connection.isPresent()) {
+            connection.get().setAutoCommit(autoCommit);
```
Comment (dagguh) [inline src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java:60]: Can't use `Optional.ifPresent` due to `throws SQLException`
Checked exceptions trying to help again 😂

**Pull request merged**